text/3513-gen-blocks.md

```diff
@@ -0,0 +1,803 @@
+
+The CLU language opened an iterator context with the `iter` keyword and produced values with `yield` statements.  E.g.:
+
+```
+odds = iter () yields (int)
+  x: int := 1
+  while x <= 20 do
+    yield x
+    x := x + 2
+  end
+end odds
+```
+
+[alphard]: https://web.archive.org/web/20150926014020/http://repository.cmu.edu/cgi/viewcontent.cgi?article=1868&context=isr
+[clu-history]: https://web.archive.org/web/20030917041834/http://www.lcs.mit.edu/publications/pubs/pdf/MIT-LCS-TR-561.pdf
+[clu-ref]: https://web.archive.org/web/20211105171453/https://pmg.csail.mit.edu/ftp.lcs.mit.edu/pub/pclu/CLU/3.Documents/MIT-LCS-TR-225.pdf
+
+## Icon
+
+In [Icon][icon-language] (introduced circa 1977), generators are woven deeply into the language, and any function can return a sequence of values.  When done explicitly, the `suspend` keyword is used.  E.g.:
+
+```
+procedure range(i, j)
+  while i < j do {
+    suspend i
+    i +:= 1
+  }
+  fail
+end
+```
+
+[icon-language]: https://web.archive.org/web/20230721102710/https://www2.cs.arizona.edu/icon/ftp/doc/lb1up.pdf
+
+## Python
+
+In Python, any function that contains a `yield` statement returns a generator.  E.g.:
+
+```python
+def odd_dup(xs):
+  for x in xs:
+    if x % 2 == 1:
+      yield x * 2
+```
+
+## ECMAScript / JavaScript
+
+In JavaScript, `yield` can be used within [`function*`][javascript-function*] generator functions.  E.g.:
+
+```javascript
+function* oddDupUntilNegative(xs) {
+  for (const x of xs) {
+    if (x < 0) {
+      return;
+    } else if (x % 2 == 1) {
+      yield x * 2;
+    }
+  }
+}
+```
+
+These generator functions are general coroutines.  `yield` forms an expression that returns the value passed to `next`.  E.g.:
+
+```javascript
+function* dup(x) {
+  while (true) {
+    x = yield x * 2;
+  }
+}
+
+const g = dup(2);
+console.assert(g.next().value === 4);
+console.assert(g.next(3).value === 6);
+```
+
+[javascript-function*]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*
+
+## Ruby
+
+In Ruby, `yield` can be used with the [`Enumerator`][ruby-enumerator] class to implement an iterator.  E.g.:
+
+```ruby
+def odd_dup_until_negative xs
+  Enumerator.new do |y|
+    xs.each do |x|
+      if x < 0
+        return
+      elsif x % 2 == 1
+        y.yield x * 2
+      end
+    end
+  end
+end
+```
+
+Ruby also uses `yield` for a general coroutine mechanism with the [`Fiber`][ruby-fiber] class.  E.g.:
+
+```ruby
+def dup
+  Fiber.new do |x|
+    while true
+      x = Fiber.yield x * 2
+    end
+  end
+end
+
+g = dup
+4 == (g.resume 2)
+6 == (g.resume 3)
+```
+
+[ruby-enumerator]: https://ruby-doc.org/3.2.2/Enumerator.html
+[ruby-fiber]: https://ruby-doc.org/3.2.2/Fiber.html
+
+## Kotlin
```

Comment:
Will note that the yield points for Kotlin's "coroutines" under the hood are delimited by continuations. Its `suspendCoroutine { coroutine -> ... }` isn't far off of `await new Promise((resolve, reject) => { ... })` in ECMAScript/JavaScript. Likewise, the end accepts a continuation object (it's an interface) to essentially act like ECMAScript's/JavaScript's `promise.then(...)`.

- Its `coroutine.resume(value)` is like ECMAScript's `resolve(value)`.
- `coroutine.resumeWithException(value)` is almost a carbon copy clone of `reject(value)`.

The way `sequence` under the hood handles it is by executing the `suspend` function with a context that store both the outer and inner continuation objects. So in essence, it's little more than a push-based `async`/`await` in the end.